pkg/resolvers/k8s/streamer.go

```diff
@@ -33,79 +41,88 @@ func startWatchingEndpointsChanges(
 			_, _ = ioutil.ReadAll(stream)
 			err = stream.Close()
 			if err != nil {
-				logrus.WithError(err).Warn("k8sresolver: Failed to Close cancelled stream connection")
+				logrus.WithError(err).Warn("Failed to Close cancelled stream connection")
 			}
 		}
 	}()
-
+	s := &streamer{
+		changeCh: make(chan change),
+		errCh:    make(chan error, 1),
+	}
 	go func() {
-		proxyAllEvents(innerCtx, json.NewDecoder(stream), eventsCh)
-		innerCancel()
+		defer innerCancel()
+
+		if err := proxy(innerCtx, json.NewDecoder(stream), s.changeCh); err != nil {
+			s.errCh <- err
+		}
 	}()
 
-	return nil
+	return s, nil
 }
 
-type eventType string
-
-const (
-	added    eventType = "ADDED"
-	modified eventType = "MODIFIED"
-	deleted  eventType = "DELETED"
-	failed   eventType = "ERROR"
-)
+func (s *streamer) ResultChans() (<-chan change, <-chan error) {
+	return s.changeCh, s.errCh
+}
 
-// event represents a single event to a watched resource.
+// We don't want to use special, internal decoder, so we need to have all typed.
 type event struct {
-	Type   eventType `json:"type"`
-	Object endpoints `json:"object"`
+	Type   watch.EventType
+	Object eventObject
+}
+
+type eventObject struct {
```

Comment:
Shouldn't these fields have JSON tags as well?

Reply:
no, they not need.

Reply:
you're not, right, sorry my bad